### PR TITLE
magma: fix #220343 and separate CUDA build/run-time dependencies

### LINKED ISSUE #220343: magma: lexicographic sorting of CUDA capabilities is not future-proof
### Describe the bug

Per https://github.com/NixOS/nixpkgs/pull/218265#discussion_r1125236419, assuming NVIDIA continues with its current compute capability naming scheme, `magma` will run into issues with the launch of Blackwell.

### Additional context

This is a followup from https://github.com/NixOS/nixpkgs/pull/218265.

### Notify maintainers

@NixOS/cuda-maintainers

### PR DESCRIPTION
**Pull request by ConnorBaker**
###### Description of changes

Goal is to shrink the size of the closure, or modularize the components which are downloaded.

Also fixes #220343.

My hope is that, for packages which do not fully support CUDA redistributables, by creating separate symlinked copies of what cudatoolkit would provide we can improve the user experience.

Closure before:

```console
/nix/store/qmnr18aqd08zdkhka695ici96k6nzirv-libunistring-1.0         	   1.7M	   1.7M
/nix/store/vv6rlzln7vhxk519rdsrzmhhlpyb5q2m-libidn2-2.3.2            	 254.1K	   2.0M
/nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224           	  28.9M	  30.8M
/nix/store/205vsmbfhq1q2vhgskpqyymqvba4mscp-gcc-11.3.0-lib           	   7.5M	  38.3M
/nix/store/2w4k8nvdyiggz717ygbbxchpnxrqc6y9-gcc-12.2.0-lib           	   7.8M	  38.6M
/nix/store/x6nnam5hk44mljbk782rcbd92jlnz8r6-pcre-8.45                	 514.4K	  31.3M
/nix/store/4vkv3rzky44hp2b8r13d8hr4ykvqhvwh-gnugrep-3.7              	 773.2K	  32.1M
/nix/store/5ynbf6wszmggr0abwifdagrixgnya5vy-bash-5.2-p15             	   1.6M	  32.4M
/nix/store/mg9l7phyhvi16p9g8g3g8fbyj4mr79gq-zlib-1.2.13              	 125.6K	  31.0M
/nix/store/6g6d4la7xsizvr8qg91f56jiqx149iqq-binutils-2.40-lib        	   2.7M	  33.7M
/nix/store/izdfacm4wfmfjbgl1k3qlfjslkkrb2kg-gfortran-12.2.0-lib      	  10.7M	  41.6M
/nix/store/yzbam1rxs24z2apzpdpgqwi5fcwbid6z-openblas-0.3.21          	  26.1M	  67.7M
/nix/store/azvwwb4994qm7b1fcz8gc3gnlf8zwzi4-blas-3                   	  53.3M	 121.0M
/nix/store/ia34rbsa6d2dalzk5f7hy5jp9zazpv24-gmp-with-cxx-6.2.1       	 729.2K	  39.3M
/nix/store/b8wwwcng8c0snvmcawvhiny4b2gr6yhh-mpfr-4.2.0               	 774.0K	  40.1M
/nix/store/jn9kg98dsaajx4mh95rb9r5rf2idglqh-attr-2.5.1               	  78.8K	  30.9M
/nix/store/bw9s084fzmb5h40x98mfry25blj4cr9r-acl-2.3.1                	 108.9K	  31.0M
/nix/store/jvl8dr21nrwhqywwxcl8di4j55765gvy-gmp-with-cxx-stage4-6.2.1	 730.4K	  39.3M
/nix/store/bg8f47vihykgqcgblxkfk9sbvc4dnksa-coreutils-9.1            	   1.4M	  40.9M
/nix/store/lp8qrhb6hs42jwbapzq20l05jf4kyicq-glibc-2.35-224-bin       	   3.0M	  33.8M
/nix/store/wyzaa007anaxxhmrcbffm597v14n2mxs-linux-headers-6.1        	   6.0M	   6.0M
/nix/store/pqnd39aq2sksad2zvswjcpkqdc7ig3f9-glibc-2.35-224-dev       	   2.2M	  42.0M
/nix/store/fhzz4yrdy17czwc9i4swhlpcp445inzb-binutils-2.40            	  28.2M	  69.6M
/nix/store/xnrwkaidhxjsb70c2bnrl6j0mmr0y3qg-expand-response-params   	  16.4K	  30.9M
/nix/store/qkshx57xqsr4g4v8ga0jn8jrgnyaam3f-binutils-wrapper-2.40    	  48.5K	  84.8M
/nix/store/f6172g5agk13134pdx5hl1qjmw8a4sdw-libmpc-1.3.1             	 273.4K	  40.4M
/nix/store/fdfcva0q4zgnm1gpc7wmz4cgq7c3hxx1-isl-0.20                 	   2.5M	  41.8M
/nix/store/vns64fmwpqz8himgfdy8h7i29h9gbryc-gcc-11.3.0               	 180.6M	 242.2M
/nix/store/ds6ivg31k3l0pjhhf3s769bkpmafa54g-gcc-wrapper-11.3.0       	  55.2K	 278.4M
/nix/store/l0d6c6wa5sk4wx70x77k3d5clmny7sw6-cuda_cudart-11.7.60      	   6.2M	   6.2M
/nix/store/mxcbimi0gs021wwyfik63km6v26pzhp8-libcublas-11.10.1.25     	   1.1G	   1.2G
/nix/store/ss8jw175bhs3cmlzmmhsbpqihy457ids-libcusparse-11.7.3.50    	 488.1M	 488.1M
/nix/store/rbm214ji7jqk58g8shhkmhvl6xflblz0-cuda_cupti-11.7.50       	  88.4M	  88.4M
/nix/store/virnlrgvysrn1nl6bra798fy6x731nj0-cuda_nvprof-11.7.50      	   9.9M	 129.1M
/nix/store/zrpivbb39kiyij6dns4pz4wh8nik6m3y-cuda_nvcc-11.7.64        	 114.8M	 393.2M
/nix/store/mpj8ij172g466nncmnfqpz694pa746dr-cuda-redist-11.7         	  63.0K	   2.1G
/nix/store/ib9ckb796i7nqj0fy6iankvhyfc6az70-magma-2.7.1              	 429.4M	   2.6G
```

Closure after:

```console
/nix/store/qmnr18aqd08zdkhka695ici96k6nzirv-libunistring-1.0         	   1.7M	   1.7M
/nix/store/vv6rlzln7vhxk519rdsrzmhhlpyb5q2m-libidn2-2.3.2            	 254.1K	   2.0M
/nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224           	  28.9M	  30.8M
/nix/store/205vsmbfhq1q2vhgskpqyymqvba4mscp-gcc-11.3.0-lib           	   7.5M	  38.3M
/nix/store/mxcbimi0gs021wwyfik63km6v26pzhp8-libcublas-11.10.1.25     	   1.1G	   1.2G
/nix/store/ss8jw175bhs3cmlzmmhsbpqihy457ids-libcusparse-11.7.3.50    	 488.1M	 488.1M
/nix/store/23ygxbxapna1d80v447mnpfa2n78ay0x-cuda-redist-11.7         	  10.5K	   1.6G
/nix/store/2w4k8nvdyiggz717ygbbxchpnxrqc6y9-gcc-12.2.0-lib           	   7.8M	  38.6M
/nix/store/x6nnam5hk44mljbk782rcbd92jlnz8r6-pcre-8.45                	 514.4K	  31.3M
/nix/store/4vkv3rzky44hp2b8r13d8hr4ykvqhvwh-gnugrep-3.7              	 773.2K	  32.1M
/nix/store/5ynbf6wszmggr0abwifdagrixgnya5vy-bash-5.2-p15             	   1.6M	  32.4M
/nix/store/mg9l7phyhvi16p9g8g3g8fbyj4mr79gq-zlib-1.2.13              	 125.6K	  31.0M
/nix/store/6g6d4la7xsizvr8qg91f56jiqx149iqq-binutils-2.40-lib        	   2.7M	  33.7M
/nix/store/izdfacm4wfmfjbgl1k3qlfjslkkrb2kg-gfortran-12.2.0-lib      	  10.7M	  41.6M
/nix/store/yzbam1rxs24z2apzpdpgqwi5fcwbid6z-openblas-0.3.21          	  26.1M	  67.7M
/nix/store/azvwwb4994qm7b1fcz8gc3gnlf8zwzi4-blas-3                   	  53.3M	 121.0M
/nix/store/ia34rbsa6d2dalzk5f7hy5jp9zazpv24-gmp-with-cxx-6.2.1       	 729.2K	  39.3M
/nix/store/b8wwwcng8c0snvmcawvhiny4b2gr6yhh-mpfr-4.2.0               	 774.0K	  40.1M
/nix/store/jn9kg98dsaajx4mh95rb9r5rf2idglqh-attr-2.5.1               	  78.8K	  30.9M
/nix/store/bw9s084fzmb5h40x98mfry25blj4cr9r-acl-2.3.1                	 108.9K	  31.0M
/nix/store/jvl8dr21nrwhqywwxcl8di4j55765gvy-gmp-with-cxx-stage4-6.2.1	 730.4K	  39.3M
/nix/store/bg8f47vihykgqcgblxkfk9sbvc4dnksa-coreutils-9.1            	   1.4M	  40.9M
/nix/store/lp8qrhb6hs42jwbapzq20l05jf4kyicq-glibc-2.35-224-bin       	   3.0M	  33.8M
/nix/store/wyzaa007anaxxhmrcbffm597v14n2mxs-linux-headers-6.1        	   6.0M	   6.0M
/nix/store/pqnd39aq2sksad2zvswjcpkqdc7ig3f9-glibc-2.35-224-dev       	   2.2M	  42.0M
/nix/store/fhzz4yrdy17czwc9i4swhlpcp445inzb-binutils-2.40            	  28.2M	  69.6M
/nix/store/xnrwkaidhxjsb70c2bnrl6j0mmr0y3qg-expand-response-params   	  16.4K	  30.9M
/nix/store/qkshx57xqsr4g4v8ga0jn8jrgnyaam3f-binutils-wrapper-2.40    	  48.5K	  84.8M
/nix/store/f6172g5agk13134pdx5hl1qjmw8a4sdw-libmpc-1.3.1             	 273.4K	  40.4M
/nix/store/fdfcva0q4zgnm1gpc7wmz4cgq7c3hxx1-isl-0.20                 	   2.5M	  41.8M
/nix/store/vns64fmwpqz8himgfdy8h7i29h9gbryc-gcc-11.3.0               	 180.6M	 242.2M
/nix/store/ds6ivg31k3l0pjhhf3s769bkpmafa54g-gcc-wrapper-11.3.0       	  55.2K	 278.4M
/nix/store/l0d6c6wa5sk4wx70x77k3d5clmny7sw6-cuda_cudart-11.7.60      	   6.2M	   6.2M
/nix/store/rbm214ji7jqk58g8shhkmhvl6xflblz0-cuda_cupti-11.7.50       	  88.4M	  88.4M
/nix/store/virnlrgvysrn1nl6bra798fy6x731nj0-cuda_nvprof-11.7.50      	   9.9M	 129.1M
/nix/store/zrpivbb39kiyij6dns4pz4wh8nik6m3y-cuda_nvcc-11.7.64        	 114.8M	 393.2M
/nix/store/fmv2m75gvby3jg106ibs4k956yl59b66-cuda-native-redist-11.7  	  53.3K	 497.8M
/nix/store/zwjfybn9vah5vkxmxl9qlawxmx7zmagp-magma-2.7.1              	 429.4M	   2.6G
```

<details>

Common derivations:

```console
/nix/store/qmnr18aqd08zdkhka695ici96k6nzirv-libunistring-1.0         	   1.7M	   1.7M
/nix/store/vv6rlzln7vhxk519rdsrzmhhlpyb5q2m-libidn2-2.3.2            	 254.1K	   2.0M
/nix/store/76l4v99sk83ylfwkz8wmwrm4s8h73rhd-glibc-2.35-224           	  28.9M	  30.8M
/nix/store/205vsmbfhq1q2vhgskpqyymqvba4mscp-gcc-11.3.0-lib           	   7.5M	  38.3M
/nix/store/2w4k8nvdyiggz717ygbbxchpnxrqc6y9-gcc-12.2.0-lib           	   7.8M	  38.6M
/nix/store/x6nnam5hk44mljbk782rcbd92jlnz8r6-pcre-8.45                	 514.4K	  31.3M
/nix/store/4vkv3rzky44hp2b8r13d8hr4ykvqhvwh-gnugrep-3.7              	 773.2K	  32.1M
/nix/store/5ynbf6wszmggr0abwifdagrixgnya5vy-bash-5.2-p15             	   1.6M	  32.4M
/nix/store/mg9l7phyhvi16p9g8g3g8fbyj4mr79gq-zlib-1.2.13              	 125.6K	  31.0M
/nix/store/6g6d4la7xsizvr8qg91f56jiqx149iqq-binutils-2.40-lib        	   2.7M	  33.7M
/nix/store/izdfacm4wfmfjbgl1k3qlfjslkkrb2kg-gfortran-12.2.0-lib      	  10.7M	  41.6M
/nix/store/yzbam1rxs24z2apzpdpgqwi5fcwbid6z-openblas-0.3.21          	  26.1M	  67.7M
/nix/store/azvwwb4994qm7b1fcz8gc3gnlf8zwzi4-blas-3                   	  53.3M	 121.0M
/nix/store/ia34rbsa6d2dalzk5f7hy5jp9zazpv24-gmp-with-cxx-6.2.1       	 729.2K	  39.3M
/nix/store/b8wwwcng8c0snvmcawvhiny4b2gr6yhh-mpfr-4.2.0               	 774.0K	  40.1M
/nix/store/jn9kg98dsaajx4mh95rb9r5rf2idglqh-attr-2.5.1               	  78.8K	  30.9M
/nix/store/bw9s084fzmb5h40x98mfry25blj4cr9r-acl-2.3.1                	 108.9K	  31.0M
/nix/store/jvl8dr21nrwhqywwxcl8di4j55765gvy-gmp-with-cxx-stage4-6.2.1	 730.4K	  39.3M
/nix/store/bg8f47vihykgqcgblxkfk9sbvc4dnksa-coreutils-9.1            	   1.4M	  40.9M
/nix/store/lp8qrhb6hs42jwbapzq20l05jf4kyicq-glibc-2.35-224-bin       	   3.0M	  33.8M
/nix/store/wyzaa007anaxxhmrcbffm597v14n2mxs-linux-headers-6.1        	   6.0M	   6.0M
/nix/store/pqnd39aq2sksad2zvswjcpkqdc7ig3f9-glibc-2.35-224-dev       	   2.2M	  42.0M
/nix/store/fhzz4yrdy17czwc9i4swhlpcp445inzb-binutils-2.40            	  28.2M	  69.6M
/nix/store/xnrwkaidhxjsb70c2bnrl6j0mmr0y3qg-expand-response-params   	  16.4K	  30.9M
/nix/store/qkshx57xqsr4g4v8ga0jn8jrgnyaam3f-binutils-wrapper-2.40    	  48.5K	  84.8M
/nix/store/f6172g5agk13134pdx5hl1qjmw8a4sdw-libmpc-1.3.1             	 273.4K	  40.4M
/nix/store/fdfcva0q4zgnm1gpc7wmz4cgq7c3hxx1-isl-0.20                 	   2.5M	  41.8M
/nix/store/vns64fmwpqz8himgfdy8h7i29h9gbryc-gcc-11.3.0               	 180.6M	 242.2M
/nix/store/ds6ivg31k3l0pjhhf3s769bkpmafa54g-gcc-wrapper-11.3.0       	  55.2K	 278.4M
/nix/store/l0d6c6wa5sk4wx70x77k3d5clmny7sw6-cuda_cudart-11.7.60      	   6.2M	   6.2M
/nix/store/mxcbimi0gs021wwyfik63km6v26pzhp8-libcublas-11.10.1.25     	   1.1G	   1.2G
/nix/store/ss8jw175bhs3cmlzmmhsbpqihy457ids-libcusparse-11.7.3.50    	 488.1M	 488.1M
/nix/store/rbm214ji7jqk58g8shhkmhvl6xflblz0-cuda_cupti-11.7.50       	  88.4M	  88.4M
/nix/store/virnlrgvysrn1nl6bra798fy6x731nj0-cuda_nvprof-11.7.50      	   9.9M	 129.1M
/nix/store/zrpivbb39kiyij6dns4pz4wh8nik6m3y-cuda_nvcc-11.7.64        	 114.8M	 393.2M
```

Derivations unique to before:

```console
/nix/store/mpj8ij172g466nncmnfqpz694pa746dr-cuda-redist-11.7         	  63.0K	   2.1G
/nix/store/ib9ckb796i7nqj0fy6iankvhyfc6az70-magma-2.7.1              	 429.4M	   2.6G
```

Derivations unique to after:

```console
/nix/store/23ygxbxapna1d80v447mnpfa2n78ay0x-cuda-redist-11.7         	  10.5K	   1.6G
/nix/store/fmv2m75gvby3jg106ibs4k956yl59b66-cuda-native-redist-11.7  	  53.3K	 497.8M
/nix/store/zwjfybn9vah5vkxmxl9qlawxmx7zmagp-magma-2.7.1              	 429.4M	   2.6G
```

</details>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ConnorBaker** [comment]
@NixOS/cuda-maintainers

File: pkgs/development/libraries/science/math/magma/generic.nix [hunk 1]
```diff
@@ -119,15 +124,14 @@ stdenv.mkDerivation {
     cmake
     ninja
     gfortran
-  ];
+  ] ++ lists.optional cudaSupport cuda-native-redist;
```
Comment (samuela) [inline pkgs/development/libraries/science/math/magma/generic.nix:127]: please prefer optionals. See https://discourse.nixos.org/t/lib-optional-considered-harmful/7081 for more info. The fact that optional exists at all is a bit unfortunate.

File: pkgs/development/libraries/science/math/magma/generic.nix [hunk 2]
```diff
@@ -119,15 +124,14 @@ stdenv.mkDerivation {
...
 
   buildInputs = [
     libpthreadstubs
     lapack
     blas
-  ] ++ lists.optionals cudaSupport [
-    cuda_joined
-  ] ++ lists.optionals rocmSupport [
+  ] ++ lists.optional cudaSupport cuda-redist
```
Comment (samuela) [inline pkgs/development/libraries/science/math/magma/generic.nix:133]: please prefer optionals. See https://discourse.nixos.org/t/lib-optional-considered-harmful/7081 for more info. The fact that optional exists at all is a bit unfortunate.

Comment (samuela) [reply pkgs/development/libraries/science/math/magma/generic.nix:133]: also let's try to minimize our impact on the git blame by only touching line 129 if possible

File: pkgs/development/libraries/science/math/magma/generic.nix [hunk 3]
```diff
@@ -80,27 +80,32 @@ let
   cudaArchitecturesString = strings.concatStringsSep ";" cudaArchitectures;
   minArch =
     let
-      minArch' = builtins.head (builtins.sort builtins.lessThan cudaArchitectures);
+      minArch' = builtins.head (builtins.sort strings.versionOlder cudaArchitectures);
     in
-    # If this fails some day, something must've changed and we should re-validate our assumptions
-    assert builtins.stringLength minArch' == 2;
```
Comment (samuela) [inline pkgs/development/libraries/science/math/magma/generic.nix:86]: why was this removed?

**Review by samuela — CHANGES_REQUESTED** [review]

File: pkgs/development/libraries/science/math/magma/generic.nix (hunk 3 above)
Comment (ConnorBaker) [reply pkgs/development/libraries/science/math/magma/generic.nix:86]: Because the assumption is incorrect, assuming NVIDIA keeps going with the same conventions they've used for the last decade. Come Blackwell, we should see the arrival of compute capability 10.0, which would translate to a `minArch'` of `"100"`, which is longer than 2.

I don't see the check providing much value, though if you'd like I can change the check to `>= 2` instead of `== 2`.

**Comment by ConnorBaker** [comment]
@SomeoneSerge I *just* understood what you meant now when you asked whether setting `CUDAHOSTCXX` was still needed. You're right, it's not necessary since you're setting it here: https://github.com/NixOS/nixpkgs/blob/a59944f1c11a5e1a6bd33cad379dae257d4bf40b/pkgs/development/compilers/cudatoolkit/redist/overrides.nix#L50.

Build fines without it, too: no more errors when trying to detect the CUDA version during CMake's configurePhase.

File: pkgs/development/libraries/science/math/magma/generic.nix (hunk 3 above)
Comment (samuela) [reply pkgs/development/libraries/science/math/magma/generic.nix:86]: gotcha gotcha, makes sense!

**Review by samuela — APPROVED** [review]

File: pkgs/development/libraries/science/math/magma/generic.nix [hunk 4]
```diff
@@ -148,13 +157,6 @@ stdenv.mkDerivation {
     "-DMAGMA_ENABLE_HIP=ON"
   ];
 
-  # NOTE: The stdenv's CXX is used when compiling the CMake test to determine the version of
-  #   CUDA available. This isn't necessarily the same as cudatoolkit.cc, so we must set
-  #   CUDAHOSTCXX.
-  preConfigure = strings.optionalString cudaSupport ''
-    export CUDAHOSTCXX=${cudatoolkit.cc}/bin/c++
-  '';
-
```
Comment (ConnorBaker) [inline pkgs/development/libraries/science/math/magma/generic.nix:157]: Removed because it's now set in `cuda_nvcc`: https://github.com/NixOS/nixpkgs/blob/a59944f1c11a5e1a6bd33cad379dae257d4bf40b/pkgs/development/compilers/cudatoolkit/redist/overrides.nix#L52 and https://github.com/NixOS/nixpkgs/blob/a59944f1c11a5e1a6bd33cad379dae257d4bf40b/pkgs/development/compilers/cudatoolkit/common.nix#L163

**Comment by ConnorBaker** [comment]
@tbenst this should be fully ready for you to review now, sorry about the delay!

File: pkgs/development/libraries/science/math/magma/generic.nix [hunk 5]
```diff
@@ -119,14 +126,16 @@ stdenv.mkDerivation {
...
+    cuda-native-redist
   ];
 
   buildInputs = [
     libpthreadstubs
     lapack
     blas
   ] ++ lists.optionals cudaSupport [
-    cuda_joined
+    cuda-redist
```
Comment (SomeoneSerge) [inline pkgs/development/libraries/science/math/magma/generic.nix:138]: Do we drop support for `cudaPackages_10`?

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 220366 --extra-nixpkgs-config '{ cudaCapabilities = [ "8.6" ]; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>magma (magma-cuda ,magma_2_7_1)</li>
    <li>magma-hip</li>
    <li>magma_2_6_2</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python310Packages.torchWithCuda.dev</li>
    <li>python310Packages.torchWithCuda.dist</li>
    <li>python310Packages.torchWithCuda.lib</li>
    <li>python310Packages.torchWithRocm</li>
    <li>python310Packages.torchWithRocm.dev</li>
    <li>python310Packages.torchWithRocm.dist</li>
    <li>python310Packages.torchWithRocm.lib</li>
  </ul>
</details>

File: pkgs/development/libraries/science/math/magma/generic.nix (hunk 5 above)
Comment (ConnorBaker) [reply pkgs/development/libraries/science/math/magma/generic.nix:138]: Hmm... I realize that because I'm using `cudaPackages`, this won't work with anything prior to CUDA 11.4, since that's when we got redistributables, right? In that case, I guess I accidentally dropped support in https://github.com/NixOS/nixpkgs/pull/217410.

Thoughts? Magma can (does?) support CUDA 10, but the only consumer in Nixpkgs is PyTorch, and they dropped support for CUDA 10 in their latest release.

Comment (SomeoneSerge) [reply pkgs/development/libraries/science/math/magma/generic.nix:138]: > won't work with anything prior to CUDA 11.4

Yes

> dropped support in ...

Oh, indeed. Maybe it's just time to prune and drop a notice in release notes. We can keep `cudaPackages_10` around for legacy packages that won't work without it, but when a package supports redist, let's default to only supporting redist. Thoughts? CC @samuela 

<details>
at least right now my feeling is that it's better to support well the things we actually use, than do a poor job at stretching for things we don't even know users for
</details>

Comment (samuela) [reply pkgs/development/libraries/science/math/magma/generic.nix:138]: > Oh, indeed. Maybe it's just time to prune and drop a notice in release notes. We can keep cudaPackages_10 around for legacy packages that won't work without it, but when a package supports redist, let's default to only supporting redist. Thoughts? CC @samuela

Agreed, we can't let these outdated versions hold us back from progress!

**Comment by samuela** [comment]
cc @tbenst friendly ping as maintainer

**Comment by tbenst** [comment]
> cc @tbenst friendly ping as maintainer

Nice work on this! I’m hoping to finish my PhD soon(TM) and had to step away from Nix in favor of productivity so I’m not in a position to review / maintain right now. Sorry about that! Please feel free to remove me as a maintainer.

**Comment by ConnorBaker** [comment]
I removed @tbenst and added myself as a maintainer.

@samuela are we okay to merge?

**Review by samuela — APPROVED** [review]
LGTM! @SomeoneSerge does this look ok to you?

**Comment by samuela** [comment]
> Nice work on this! I’m hoping to finish my PhD soon(TM) and had to step away from Nix in favor of productivity so I’m not in a position to review / maintain right now. Sorry about that! Please feel free to remove me as a maintainer.

No problem, @tbenst . Best of luck with your defense!

**Comment by SomeoneSerge** [comment]
Result of `nixpkgs-review pr 220366 --extra-nixpkgs-config '{ cudaCapabilities = [ "8.6" ]; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>11 packages built:</summary>
  <ul>
    <li>magma (magma-cuda ,magma_2_7_1)</li>
    <li>magma-hip</li>
    <li>magma_2_6_2</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python310Packages.torchWithCuda.dev</li>
    <li>python310Packages.torchWithCuda.dist</li>
    <li>python310Packages.torchWithCuda.lib</li>
    <li>python310Packages.torchWithRocm</li>
    <li>python310Packages.torchWithRocm.dev</li>
    <li>python310Packages.torchWithRocm.dist</li>
    <li>python310Packages.torchWithRocm.lib</li>
  </ul>
</details>

**Review by SomeoneSerge — APPROVED** [review]

**Comment by samuela** [comment]
Great, thanks so much @ConnorBaker and @SomeoneSerge !
